### ISSUE TITLE
Housekeeping

### ISSUE THREAD

**Issue by mstormi** [openhab/openhabian#1446]
Had sort of an outage today, systemd / journald were using 100% CPU because `/var/log` on ZRAM had filled 100%.

I did below to fix. Trouble is, I would have expected Raspi OS to do that.
/lib/systemd/system/systemd-tmpfiles-clean.service execute /bin/systemd-tmpfiles --clean **daily** but I think it does not handle journal.
We should setup a systemd timer to a) run logrotate -f /etc/logrotate.conf and b) clean the journal at regular intervals or if fill % of var/log is above some level.

Double /var/log size in /etc/ztab

I think the majority of log entries started to be generated after openHAB was unable to write to disk otherwise it would have produced way less data.  Still we have the potential problem e.g. If some beginner finds out about "debug all" then forgets to turn off he's in trouble quick.
Cross-check default OH3 log(size + number of) setting, usually it's a user responsibility but maybe we should restrict that to match/correlate with ztab setting ?

```
[09:05:49] root@openhab:/home/openhabian# journalctl --vacuum-time=7d
Deleted empty archived journal /var/log/journal/b51c418b6d5c40d0a56e24cd7d89369b/system@0005ba41c3d3d509-f89920cc0274a797.journal~ (7.5M).
Deleted empty archived journal /var/log/journal/b51c418b6d5c40d0a56e24cd7d89369b/system@0005ba3dbcd1bba9-f573024c102a525b.journal~ (7.9M).
Deleted empty archived journal /var/log/journal/b51c418b6d5c40d0a56e24cd7d89369b/system@0005ba4161f8aea0-37974c6259d0137f.journal~ (7.9M).
Deleted empty archived journal /var/log/journal/b51c418b6d5c40d0a56e24cd7d89369b/system@0005ba4040ad855f-7baf052488450af8.journal~ (7.9M).
Deleted empty archived journal /var/log/journal/b51c418b6d5c40d0a56e24cd7d89369b/system@0005ba3eeaa33e29-45f151f554e28b43.journal~ (7.3M).
Deleted empty archived journal /var/log/journal/b51c418b6d5c40d0a56e24cd7d89369b/system@0005ba3c42122a54-7d70cf19b73011b0.journal~ (6.2M).
Vacuuming done, freed 45.0M of archived journals from /var/log/journal/b51c418b6d5c40d0a56e24cd7d89369b.
Vacuuming done, freed 0B of archived journals from /run/log/journal/b51c418b6d5c40d0a56e24cd7d89369b.
[09:06:12] root@openhab:/home/openhabian# 
```

**Comment by ecdye** [comment]
I don't think we want to make the ztab size any bigger, it has been fine up to this point. More likely we need to adjust the log rotation.

**Comment by mstormi** [comment]
ztab increase should not do harm as it is the *uncompressed* fs limit that gets hit not the (RAM limited) mem limit.
System log rotation isn't the deal, it's totally dependent on OH logging so that we must start tweaking.

**Comment by ecdye** [comment]
> ztab increase should not do harm as it is the _uncompressed_ fs limit that gets hit not the (RAM limiting) mem limit.

I changed it slightly, it should default to 300M mem limit now, but there is overhead and it wastes a little bit of ram for every bit that we add, I have observed it and on some systems it could cause issues if we go too high.

I wouldn't hurt to look into the way that OH logs but it also has not been an issue before so I doubt that it will be very high priority.

**Comment by mstormi** [comment]
I didn't mean to increase the memlimit but the disk limit.
Given I see a compression rate of ~3,5:1 what about mem 200M/disk 750M ?

**Comment by ecdye** [comment]
It shouldn't ever need to be that high as it the compression in only so good, currently log it set to `200M` for mem limit and `600M` Disk limit. This is a good middle ground as it is 3 times more than the mem limit, I don't think that your solution should be necessary in most cases, perhaps just on your box because as a dev you do more logging than a normal user.

**Comment by mstormi** [comment]
I don't think you get my point. Today 600M disk limit is limiting every user. Note 600M is far less than / has free space (which would be the limit if no ZRAM). Raising that does not harm but allows for more logs. Now say compression rate is 3,5:1 then we can write 700M to disk before ZRAM returns file write error because mem limit 200M * 3,5 = 700. With current settings we cannot, it tops out at 600M. With 750 or even 1000 it would top out at 700.
We don't harm anybody with setting disk limit high as long as we don't change the mem limit so no, there is NO need for tradeoff / finding "middle ground".
That we would only need to do if I wanted to increase mem limit to increase effective log space beyond 700M.

The bottom line is about tweaking openhab logging though so it rotates well before 700M (or 700M minus system logs) are reached.

**Comment by ecdye** [comment]
The problem is that I have never observed a compression ratio that high, even using `zstd` the best I have ever personally observed is ~3:1. Additionally, adding a higher disk limit does take some overhead in the RAM look at the documentation for `zram-config` it states in there that there is a noticeable overhead in memory for creating a large disk limit, and because I have never seen the ratio go that high I see little point in increasing the disk size to only increase the amount of RAM it uses.

**Comment by mstormi** [comment]
It depends on the data obviously. Maybe because I mostly have zwave logging where the log message 'frame' is always the same hence well compressible. This is my production system. See the ratio on zram2, that is persistence (rrd4j create nulled bytes that's why). Also see the minimal overhead i.e. difference TOTAL vs COMPR and the difference zram1 vs zram2. It also depends way more on the amount of data finally stored than on the disk limit. I guess by "large disk limit" they mean something like a couple of TB which you could think of simply putting there as well if there wasn't this hint.

openhabian@openhab:/etc/amanda $  zramctl
NAME       ALGORITHM DISKSIZE   DATA COMPR TOTAL STREAMS MOUNTPOINT
/dev/zram2 lzo-rle      1000M   318M  7.5M  9.3M       4 /opt/zram/zram2
/dev/zram1 lzo-rle       500M 199.5M 58.8M   62M       4 /opt/zram/zram1
/dev/zram0 lzo-rle       600M    98M   32M 34.5M       4 [SWAP]

**Comment by ecdye** [comment]
It is though, on my box I have tested and it results in an increase in +15M extra that never gets used, to me it seems like a waste. If you think it is really important go ahead.

**Comment by mstormi** [comment]
+15M when increasing disksize frmo 600M to 750M? Really?

**Comment by ecdye** [comment]
For me yeah

**Comment by mstormi** [comment]
are you sure it's because of the limit and not the (changed) file data ?
Remember it syncs and removes ALL data from the upper FS on changes/reboots.

**Comment by ecdye** [comment]
Yes, I'm sure. I was testing this a lot over the last weekend because I was playing with `zram-config` and was trying to see what different settings would cause.

**Comment by mstormi** [comment]
hmm ok then let's keep ztab as it is now. journal is limited to 50M so I think no point in changing that either.
Only thing left is OH logging then. If I read the new logging config /var/lib/openhab/etc/log4j2.xml right, it does file chunks of 16MB but there does not seem to be a limit on the number of files (possibly there is some default ?).

**Comment by ecdye** [comment]
Another issue may have been that there was a bug present in the zram code for a little while that would have cause the `mem_limit` to also be used as the `disk_limit` but is now fixed. See <https://community.openhab.org/t/zram-issues/100552/20>

**Comment by mstormi** [comment]
what does that bug mean w.r.t. increased mem usage when we just slightly increase the disk limit, is it still the case ?

**Comment by ecdye** [comment]
Yes that is still the case. But the disk size should be properly set now and that could have potentially limited the disk size before causing the issue that you had.

**Comment by mstormi** [comment]
Yes so all is left is logfiles.
Still trying to figure out the default number of files in rotation in log4j2.

**Comment by mstormi** [comment]
Could not find really find an authoritative statement on the number of default logs that are kept on rolling over
(what I found says "1" but that obviously does not apply).
So tested and it seems to keep 7 files plus the active one, all are 16M by default so 8*16M=128M max disk usage.
In theory it's twice that (to include events.log.*) but in practice I rarely saw any events.log roll over.
Plus there's the system logs which are a little uncalculable. However as Linux/Raspi OS distribution fully takes care of limiting that (logrotate) I assume that under normal usage those will top out at what I'm seeing on my systems which is another 100-150M.
That being said, I think our current ztab setting of 200/600 is pretty much fine (enough but not way too much headroom).
So in total this is why I'll close this issue now.
